arangod/RocksDBEngine/RocksDBVPackIndex.cpp

```diff
@@ -640,88 +639,76 @@ Result RocksDBVPackIndex::insertInternal(transaction::Methods* trx,
   RocksDBValue value = _unique ? RocksDBValue::UniqueVPackIndexValue(documentId)
                                : RocksDBValue::VPackIndexValue();
 
   size_t const count = elements.size();
-  RocksDBValue existing = RocksDBValue::Empty(RocksDBEntryType::UniqueVPackIndexValue);
-  IndexingDisabler guard(mthds, !_unique && trx->hasHint(transaction::Hints::Hint::FROM_TOPLEVEL_AQL));
-
+  rocksdb::PinnableSlice existing;
   for (size_t i = 0; i < count; ++i) {
     RocksDBKey& key = elements[i];
     if (_unique) {
-      if (mthds->Exists(_cf, key)) {
-        res = TRI_ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED;
-        auto found = mthds->Get(_cf, key, existing.buffer());
-        TRI_ASSERT(found.ok());
-      }
-    }
-
-    if (res == TRI_ERROR_NO_ERROR) {
-      arangodb::Result r =
-          mthds->Put(_cf, key, value.string(), rocksutils::index);
-      if (!r.ok()) {
-        res = r.errorNumber();
+      s = mthds->Get(_cf, key.string(), &existing);
+      if (s.ok()) { // detected conflicting index entry
+        res.reset(TRI_ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED);
+        break;
       }
     }
 
-    if (res != TRI_ERROR_NO_ERROR) {
-      for (size_t j = 0; j < i; ++j) {
-        mthds->Delete(_cf, elements[j]);
-      }
-
-      if (res == TRI_ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED && !_unique) {
-        // We ignore unique_constraint violated if we are not unique
-        res = TRI_ERROR_NO_ERROR;
-        // TODO: remove this? seems dangerous...
-      }
+    s = mthds->Put(_cf, key, value.string());
+    if (!s.ok()) {
+      res = rocksutils::convertStatus(s, rocksutils::index);
       break;
     }
   }
 
-  if (res == TRI_ERROR_NO_ERROR && !_unique) {
+  if (res.ok() && !_unique) {
     auto state = RocksDBTransactionState::toState(trx);
-
     for (auto& it : hashes) {
       // The estimator is only useful if we are in a non-unique indexes
       TRI_ASSERT(!_unique);
       state->trackIndexInsert(_collection.id(), id(), it);
     }
-  }
-
-  if (res == TRI_ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED) {
-    LocalDocumentId documentId = RocksDBValue::documentId(existing);
+  } else if (res.is(TRI_ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED)) {
+    // find conflicing document
+    LocalDocumentId docId = RocksDBValue::documentId(existing);
     std::string existingKey;
-
-    bool success = _collection.getPhysical()->readDocumentWithCallback(trx, documentId, [&](LocalDocumentId const&, VPackSlice doc) {
-      existingKey = doc.get(StaticStrings::KeyString).copyString();
+    bool success = _collection.getPhysical()->readDocumentWithCallback(trx, docId,
+    [&](LocalDocumentId const&, VPackSlice doc) {
+      existingKey = transaction::helpers::extractKeyFromDocument(doc).copyString();
     });
     TRI_ASSERT(success);
 
     if (mode == OperationMode::internal) {
-      return IndexResult(res, std::move(existingKey));
+      res.resetErrorMessage(std::move(existingKey));
+    } else {
+      addErrorMsg(res);
```

Comment:
can we add the existing key to the error message here? I think it will be missing now.